packages/vite/src/node/plugins/define.ts

```diff
@@ -116,14 +129,30 @@ export function definePlugin(config: ResolvedConfig): Plugin {
         return
       }
 
-      const [define, pattern] = ssr ? ssrPattern : defaultPattern
+      const [define, pattern, undefinedPattern, banner] = ssr
+        ? ssrPattern
+        : defaultPattern
       if (!pattern) return
 
       // Check if our code needs any replacements before running esbuild
       pattern.lastIndex = 0
       if (!pattern.test(code)) return
 
-      return await replaceDefine(code, id, define, config)
+      // process undefined imports.meta.env properties
+      const defineWithUndefined = { ...define }
+      undefinedPattern.lastIndex = 0
+      for (const undefinedEnvKey of [...code.matchAll(undefinedPattern)]) {
+        defineWithUndefined[undefinedEnvKey[0]] = 'undefined'
+      }
+
+      return await replaceDefine(
+        code,
+        id,
+        defineWithUndefined,
+        config,
+        // if there is bare `import.meta.env`, then add the manual `import.meta.env`
+        bareImportMetaEnvRe.test(code) ? banner : undefined,
```

Comment:
Just a thought, but I wonder if we can improve perf and false positives on this part too. Instead of using `banner`, we could manually append the banner ourselves after `replaceDefine` returns, and to deal with sourcemaps the banner could be a single line with `\n` so that we can simply edit `map.mappings = ';' + map.mappings`. IIRC `;` means a new line which would represent the banner line.

Not suggesting for you to do this here though, but something I wonder we could squeeze in the future.